llvm/lib/Analysis/ValueTracking.cpp

```diff
@@ -8781,40 +8781,14 @@ static SelectPatternResult matchSelectPattern(CmpInst::Predicate Pred,
   return matchFastFloatClamp(Pred, CmpLHS, CmpRHS, TrueVal, FalseVal, LHS, RHS);
 }
 
-/// Helps to match a select pattern in case of a type mismatch.
-///
-/// The function processes the case when type of true and false values of a
-/// select instruction differs from type of the cmp instruction operands because
-/// of a cast instruction. The function checks if it is legal to move the cast
-/// operation after "select". If yes, it returns the new second value of
-/// "select" (with the assumption that cast is moved):
-/// 1. As operand of cast instruction when both values of "select" are same cast
-/// instructions.
-/// 2. As restored constant (by applying reverse cast operation) when the first
-/// value of the "select" is a cast operation and the second value is a
-/// constant.
-/// NOTE: We return only the new second value because the first value could be
-/// accessed as operand of cast instruction.
-static Value *lookThroughCast(CmpInst *CmpI, Value *V1, Value *V2,
-                              Instruction::CastOps *CastOp) {
+static Value *lookThroughCastConst(CmpInst *CmpI, Value *V1, Value *V2,
+                                   Instruction::CastOps *CastOp) {
```

Comment:
This still doesn't use the right types, it should be something like:
```suggestion
static Value *lookThroughCastConst(CmpInst *CmpI, CastInst *Cast1, Constant *C,
                                   Instruction::CastOps *CastOp) {
```
and then you don't need the dyn_casts in this function.